Ppa issue#50

hansvancalster opened the pull request:
closes #50

hansvancalster committed:
mention PPA and also select the name field

hansvancalster committed:
add new function get_inboveg_ppa

hansvancalster committed:
bump version

hansvancalster committed:
add NEWS item

hansvancalster committed:
fix language

hansvancalster committed:
modifications by setup_package()

hansvancalster committed:
spelling

hansvancalster committed:
allowed words custom dictionary

hansvancalster committed:
spelling

hansvancalster committed:
fix duplicate names

hansvancalster committed:
update vignette

hansvancalster committed:
fix linters

hansvancalster committed:
add keywords and community

hansvancalster committed:
update vignette data

hansvancalster committed:
update description

hansvancalster committed:
add affiliations and move dbplyr to suggests

hansvancalster committed:
update citation files

hansvancalster committed:
use dbplyr conditionally

hansvancalster committed:
move methods to suggests

hansvancalster committed:
update rda file

hansvancalster committed:
fix url

hansvancalster committed:
update rda

hansvancalster committed:
update checklist.yml

hansvancalster committed:
update SQL

hansvancalster committed:
also add comment field

hansvancalster committed:
update checklist.yml

hansvancalster committed:
update rda

hansvancalster committed:
add a new section #48

hansvancalster commented:
closes #48

hansvancalster committed:
fix example

hansvancalster committed:
also select taxongroupcode

hansvancalster commented:
closes #47

hansvancalster committed:
reindent lines

hansvancalster committed:
allow exporting additional variables

hansvancalster committed:
export comment field

hansvancalster committed:
fix lintr and update rda file

hansvancalster commented:
@ElsLommelen checklist complains about cyclocomp too high for functions `get_inboveg_ppa` and `get_inboveg_recording`. I don't want to deal with it: it works just fine. IMO abstracting away the logic (mainly if statements) into separate subfunctions will not make the code clearer. So I would suggest to leave it as is. I was not able to let checklist ignore this warning - not sure why. I already made lots of small changes suggested by checklist::check_package, as you can tell by the many files that are changed by this PR...

ElsLommelen commented:
> checklist complains about cyclocomp too high for functions `get_inboveg_ppa` and `get_inboveg_recording`. I don't want to deal with it: it works just fine. IMO abstracting away the logic (mainly if statements) into separate subfunctions will not make the code clearer.

I had a quick look at this, and I kind of understand why the cyclocomp value is so extremely high: as an outstander it is very hard to see the difference between the if-clauses: it seems to be the same each time, with at least 6 different logical expressions combined with && and (), and only after looking very closely, I noticed the ! in front of some of the logical expressions.  I don't know if it would be possible to separate these expressions and glue them together afterwards?  E.g.:
```
sql_survey_name <-
  ifelse(!(length(survey_name) == 1 && survey_name[1] == "%"), "AND ivS.Name IN ({survey_name*})", "")
# and the same for all other parts
# and finally:
sql_statement <- glue_sql(common_part, qql_survey_name, ...)
```
I don't know the details, so I'm not sure this is possible in this case, but to me this code would be much more readable, easier to maintain, shorter, and I'm sure it would pass the cyclcomp test without splitting the function (certainly when using ifelse instead of if).

ElsLommelen commented on R/get_inboveg_ppa.R:
```diff
@@ -0,0 +1,294 @@
+        user_reference = user_reference,
+        .con = connection)
+    }
+    if (!exists("sql_statement")) {
+      stop("When choosing multiple = TRUE, please add a (vector of) value(s) for at least one of the following variables: survey_name, user_reference or recording_givid.") #nolint
+    }
```
```suggestion
    stopifnot(
      "When choosing multiple = TRUE, please add a (vector of) value(s) for at least one of the following variables: survey_name, user_reference or recording_givid." #nolint
      = exists("sql_statement")
    )
```
And another trick to decrease the cyclcomp a bit.  ;-)
(`assert_that()` is also possible, but due to the `stop` and the `!`,  I spontanuously thought at `stopifnot()`)

ElsLommelen commented on inst/en_gb.dic:
```diff
@@ -0,0 +1,19 @@
+Github
```
Is Git**H**ub in the standard library?

hansvancalster replied:
[17ab200](https://github.com/inbo/inbodb/pull/51/commits/17ab200e97d0d3d067a30999fba75bf51f35478f)

ElsLommelen commented on inst/en_gb.dic:
```diff
@@ -0,0 +1,19 @@
+MILKLIM
+Natura
+PPA
+Rstudio
```
Is RStudio in the standard library?

hansvancalster replied:
[17ab200](https://github.com/inbo/inbodb/pull/51/commits/17ab200e97d0d3d067a30999fba75bf51f35478f)

ElsLommelen commented on inst/en_gb.dic:
```diff
@@ -0,0 +1,19 @@
+Sublicensing
+WIPO
+defenses
+favor
```
is it favour in British English?

hansvancalster replied:
Yes , but "favor" is in the LICENSE file if I remember well. So I did not want to change that as it will trigger an alarm from checklist. If it needs to be changed, it would need to be changed in the template license files included in the checklist package.

ElsLommelen replied:
As checklist checks the whole file, there is no need to include it in the spelling check.  Look how I solve it in [another package](https://github.com/inbo/dhcurve/blob/main/checklist.yml) to keep checklist originated packages as is while not adding 'wrong' words to the dictionary.  (Because this has the implication that these 'wrong' words can be used everywhere in the package without error.)  (I used the `set_exceptions()` function to accomplish this, and I was advised by @ThierryO to exclude the LICENSE file as it was checked by checklist anyway.)

ElsLommelen commented on R/get_inboveg_ppa.R:
```diff
@@ -0,0 +1,294 @@
+#' from a point location) for one or more surveys,
+#' or in combination with the unique ID (`recordingGIVID`) or user reference
+#'
+#' @param user_reference A character string or a character vector giving the
+#' name of a recording for which you want to extract relevé information.
+#' As default (`user_reference` = "%") all user-references are returned.
+#' @param recording_givid A character string or a character vector giving
+#' the unique id of a recording for which you want to extract relevé
+#' information.
+#' As default (`recording_givids` = "%") all `recording_givids` are returned.
+#' @param survey_name A character string or a character vector, depending on
+#' multiple parameter, giving the name or names of the survey(s) for which you
+#' want to extract relevé information. As default (`survey_name` = "%") all
+#' surveys
+#' are returned.
+#' @param connection `dbconnection` with the database 'Cydonia'
+#' on the `inbo-sql07-prd` server
+#' @param collect If FALSE (the default), a remote `tbl` object is returned.
+#' This is like a reference to the result of the query but the full result of
+#' the query is not brought into memory. If TRUE the full result of the query is
+#' collected (fetched) from the database and brought into memory of the working
+#' environment.
+#' @param multiple If TRUE survey_name, `user_reference` or `recording_givid`
+#' can
+#' take a character vector with multiple survey names, user references or
+#' `recording_givids` that must match exactly.
+#' If FALSE (the default) survey_name , `user_reference` or `recording_givid`
+#' must
+#' be a single character string (one survey name, or one user_reference or one
+#' `recording_givid`). Only survey_name can include wildcards to allow partial
+#' matches.
```
```suggestion
#' @inheritParams get_inboveg_recording
```
It seems all arguments and its descriptions are the same, please save maintenance time by referring to the descriptions instead of making a copy.

ElsLommelen commented on DESCRIPTION:
```diff
@@ -21,21 +21,23 @@ License: GPL-3
 Suggests: 
+    dbplyr,
     knitr,
+    methods,
```
I'm not sure our implementations of the methods dbDisconnect and dbFetch are still functioning correctly if users haven't installed this?  But I can have a look at this while solving issue #52.  Maybe move it back to imports for now, if possible?

hansvancalster replied:
The vignettes work, so I don't think there is a problem with `dbDisconnect`. Putting them in imports will cause a checklist error or warning that these packages are imported but not used.

ElsLommelen reviewed (commented):
Thanks a lot for your hard work!
Except for the cyclocomp issue, I found a minor issue from the maintenance point of view (and a technical one I will look at later), and I was just wondering on some of the words added to the dictionary (because I am using them differently myself).  Just have a look if they are known in the main library after changing to my suggestion, and if not, you can leave them as is (or choose yourself which you like best).

hansvancalster committed:
Update R/get_inboveg_ppa.R

Co-authored-by: Els Lommelen <els.lommelen@inbo.be>

hansvancalster committed:
Update R/get_inboveg_ppa.R

Co-authored-by: Els Lommelen <els.lommelen@inbo.be>

hansvancalster committed:
update param description

hansvancalster committed:
address dictionary comments

hansvancalster committed:
rework logic to built query

hansvancalster commented:
> I had a quick look at this, and I kind of understand why the cyclocomp value is so extremely high: as an outstander it is very hard to see the difference between the if-clauses: it seems to be the same each time, with at least 6 different logical expressions combined with && and (), and only after looking very closely, I noticed the ! in front of some of the logical expressions. I don't know if it would be possible to separate these expressions and glue them together afterwards? E.g.:
> 
> ```
> sql_survey_name <-
>   ifelse(!(length(survey_name) == 1 && survey_name[1] == "%"), "AND ivS.Name IN ({survey_name*})", "")
> # and the same for all other parts
> # and finally:
> sql_statement <- glue_sql(common_part, qql_survey_name, ...)
> ```
> 
> I don't know the details, so I'm not sure this is possible in this case, but to me this code would be much more readable, easier to maintain, shorter, and I'm sure it would pass the cyclcomp test without splitting the function (certainly when using ifelse instead of if).


Thanks for the ideas. I have completely reworked and simplified the logic behind all the if statements. The current implementation does not cover exactly the same functionallity as before:

-  wildcards with LIKE statements are now also allowed for `user_reference` and `recording_givid`
- all WHERE sub-clauses are combined with AND. Previously, in case of `multiple = TRUE` the `survey_name` was combined with (`user_reference` OR `recording_givid`). This behaviour was not document and not clear for the user.
- The `multiple` argument is no longer needed and I have removed it
- I have abstracted it in a small helper function for easier re-use in `get_inboveg_ppa` and `get_inboveg_recording`

hansvancalster committed:
fix typo

ElsLommelen commented on NEWS.md:
```diff
@@ -2,6 +2,8 @@
 * Add new function `get_inboveg_ppa()` to get point-plant-distance type data
   (PPA) from INBOVEG (#50)
+* Reworked `get_inboveg_recording()`. Argument `multiple` is no longer needed
+  and therefor removed.
```
Could you please reword this message looking at it from the user point of view: what changes for them?  Are there any breaking changes?  And we might want to keep argument multiple for a while, give it default value NA and give a warning referring to this change if the user gives it value TRUE or FALSE.

hansvancalster replied:
I think this should now be clear from the deprecate_warn() message that will be generated if a user uses the deprecated argument.

hansvancalster replied:
[0c64065](https://github.com/inbo/inbodb/pull/51/commits/0c64065e752f09e589180f32f8731bfc1a74257b)

ElsLommelen replied:
There are different types of users:
- some users follow a package of interest and read the information when a new version of a package comes out, and they want to know what has changed from their (user) point of view: what code changes are needed, what functionality has changed, new features,...?  This message is for them, but with 'reworked `get_inboveg_recording()`' they have no clue what changes to expect.  Maybe just mention that argument multiple is removed?  Or that the function is simplified and the argument multiple is removed???
- other users just install a new version when they have to (e.g. after R is updated at INBO), and after running their code they will notice if packages have changed.  They will be warned by the deprecation warning (and might find the news file when looking for more information)

So with this message in NEWS.md, you will likely reach another user group than with the deprecation warning, a group that wishes to be informed before installing a new version.  (As the majority of the users most likely belongs to the second group, I found important to cover both.  ;-) )

hansvancalster replied:
[beadfc1](https://github.com/inbo/inbodb/pull/51/commits/beadfc1226686c68520dae15254d4cdb45f07d4e)

ElsLommelen commented on R/get_inboveg_recording.R:
```diff
@@ -262,6 +167,55 @@ get_inboveg_recording <- function(
+    connection) {
+  survey_where <- ifelse(
+    length(survey_name) != 1,
+    " AND ivS.Name IN ({survey_name*})",
+    " AND ivS.Name LIKE {survey_name}"
+  )
+
+  user_reference_where <- ifelse(
+    length(user_reference) != 1,
+    "ivR.UserReference IN ({user_reference*})",
+    "ivR.UserReference LIKE {user_reference}"
+  )
+
+  recording_givid_where <- ifelse(
+    length(recording_givid) != 1,
+    "ivR.[RecordingGivid] IN ({recording_givid*})",
+    "ivR.[RecordingGivid] LIKE {recording_givid}"
+  )
+
+  combined_where_clauses <-
+    paste0(
+      survey_where,
+      " AND ",
+      user_reference_where,
+      " AND ",
+      recording_givid_where
+    )
+
+  sql_statement <- glue_sql(
+    common_part,
+    combined_where_clauses,
+    survey_name = survey_name,
+    recording_givid = recording_givid,
+    user_reference = user_reference,
+    .con = connection)
+
+  return(sql_statement)
+}
+
+
```
Could you add this function to a separate file named after the function with some explanation to make it easier for developers to find their way in the package?  When making changes in `get_inboveg_ppa()`, nobody will search for this function here.
For the documentation, just mention inboveg in the title and add explain in one sentence it is a helper function for functions xxx (with maybe some minimal explanation to understand it as a developer).  And add `@noRd` if the function is not exported.

hansvancalster replied:
[391a74f](https://github.com/inbo/inbodb/pull/51/commits/391a74fe504d1ea86507ea1b7048c2bfd534d82c)

ElsLommelen reviewed (commented):
Looks good (without having looked at every detail), but I noticed some small issues that I would like to be adressed for the ease of users and collaborators.

ElsLommelen commented:
> * all WHERE sub-clauses are combined with AND. Previously, in case of `multiple = TRUE` the `survey_name` was combined with (`user_reference` OR `recording_givid`). This behaviour was not document and not clear for the user.

I suppose replacing AND by OR implies that the use is more limited, and that the user has to run the function several times to get the same result?  Could you judge if this could be a problem for users?
I can't think of an example that would cause a problem here, but I am not familiar with the database and I don't know which information users need from this database (which makes it impossible for me to look at it from the users point of view).

hansvancalster committed:
move to separate R file

hansvancalster commented:
> > * all WHERE sub-clauses are combined with AND. Previously, in case of `multiple = TRUE` the `survey_name` was combined with (`user_reference` OR `recording_givid`). This behaviour was not document and not clear for the user.
> 
> I suppose replacing AND by OR implies that the use is more limited, and that the user has to run the function several times to get the same result? Could you judge if this could be a problem for users? I can't think of an example that would cause a problem here, but I am not familiar with the database and I don't know which information users need from this database (which makes it impossible for me to look at it from the users point of view).

Yes, it is more limiting but I think the use case does not occur. I've never seen a user use it and not one of the examples in the example section or the vignette demonstrate what happens if you pass both a specific user reference and a givid. This is already an indication that it will be unlikely. Most users will filter by survey name, possibly combined with either user references or givids. The user reference is what the user uses to "ID" her/his relevés, while the givids are an internal ID provided by INBOVEG. Only givids are unique across the database; user references are not guaranteed to be unique AFAIK and INBOVEG users will be aware of this difference.

ElsLommelen commented:
> > > * all WHERE sub-clauses are combined with AND. Previously, in case of `multiple = TRUE` the `survey_name` was combined with (`user_reference` OR `recording_givid`). This behaviour was not document and not clear for the user.
> > 
> > 
> > I suppose replacing AND by OR implies that the use is more limited, and that the user has to run the function several times to get the same result? Could you judge if this could be a problem for users? I can't think of an example that would cause a problem here, but I am not familiar with the database and I don't know which information users need from this database (which makes it impossible for me to look at it from the users point of view).
> 
> Yes, it is more limiting but I think the use case does not occur. I've never seen a user use it and not one of the examples in the example section or the vignette demonstrate what happens if you pass both a specific user reference and a givid. This is already an indication that it will be unlikely. Most users will filter by survey name, possibly combined with either user references or givids. The user reference is what the user uses to "ID" her/his relevés, while the givids are an internal ID provided by INBOVEG. Only givids are unique across the database; user references are not guaranteed to be unique AFAIK and INBOVEG users will be aware of this difference.

Ok, then we should just keep your new logic.  (OR might also lead to unexpected or unwanted behaviour if the user is not aware of it.)

hansvancalster committed:
deprecate multiple

hansvancalster committed:
update NEWS

ElsLommelen committed:
fix lintr error on 'object complete_sql_statement not found'

ElsLommelen committed:
update .zenodo.json

ElsLommelen committed:
move package methods back to 'imports'

ElsLommelen committed:
update .Rda from vignette

ElsLommelen committed:
motivate why methods is needed
